python/rainbow/simulators/proximal_contact/solver_interface.py

```diff
@@ -0,0 +1,133 @@
+            self.last_merit = self.merit
+            np.copyto(self.sol, self.x)
+
+    def initialize_stats(self, timer):
+        """ Initialize the statistics information, when debug_on is True.
+
+        Args:
+            timer ('Timer'): The timer used to record the time.
+        """
+        self.timer = timer
+        self.stats[self.prefix + "residuals"] = (
+                np.ones(self.engine.params.max_iterations, dtype=np.float64) * np.inf
+            )
+        self.stats[self.prefix + "lambda"] = np.zeros(
+                [self.engine.params.max_iterations] + list(self.b.shape), dtype=np.float64
+            )
+        self.stats[self.prefix + "reject"] = np.zeros(self.engine.params.max_iterations, dtype=bool)
+        self.stats[self.prefix + "exitcode"] = 0
+        self.stats[self.prefix + "iterations"] = self.engine.params.max_iterations
+        self.timer.start()
+
+    def check_convergence(self):
+        """ Check the convergence.
+
+        Returns:
+            bool: if True, the solver converges, otherwise(False), the solver does not converge.
+        """
+        np.subtract(self.x, self.sol, self.error)
+        self.merit = np.linalg.norm(self.error, np.inf)
+
+        if self.debug_on:
+            self.stats[self.prefix + "lambda"][self.iteration] = self.x
+            self.stats[self.prefix + "residuals"][self.iteration] = self.merit
+
+        if self.merit < self.engine.params.absolute_tolerance:
+            if self.debug_on:
+                self.stats[self.prefix + "iterations"] = self.iteration
+                self.stats[self.prefix + "exitcode"] = 1
+                self.timer.end()
+                self.stats[self.prefix + "solver_time"] = self.timer.elapsed
+            return True
+        if np.abs(self.merit - self.last_merit) < self.engine.params.relative_tolerance * self.last_merit:
+            if self.debug_on:
+                self.stats[self.prefix + "iterations"] = self.iteration
+                self.stats[self.prefix + "exitcode"] = 2
+                self.timer.end()
+                self.stats[self.prefix + "solver_time"] = self.timer.elapsed
+            return True
+        return False
+    
+    @abstractmethod
+    def sweep(self):
+        """ Sweep the contact points to compute the new contact force.
+            This method should be implemented in the derived class.
+        """
+        pass
+    
+    def solve(self):
+        """ Solve the contact problem.
+
+        Returns: The new contact force and the statistics information.
+        """
+        if self.debug_on:
+            self.initialize_stats(Timer("CONTACT_SOLVER"))
+
+        # Compute initial r-factor value
+        self.compute_initial_r_factor()
+
+        for iteration in range(self.engine.params.max_iterations):
+            self.iteration = iteration
+            self.sweep()
+            # Check convergence
+            if self.check_convergence():
+                return self.x, self.stats
```

Comment:
What happened with "sol"?

Reply:
My mistake, here should return the `sol`, but I did not catch any bugs when running soft body demos...  I will check here again.